STI foreman provider

kbrock opened the pull request:
This creates the models to be used for creating the foreman configuration provider.



This feels like it needs a lot more.
Did I miss anything?
/cc @gmcculloug @brandondunne

Fryguy commented on vmdb/app/models/computer_system.rb:
```diff
@@ -0,0 +1,6 @@
+  belongs_to :configured_system
+
+  has_one :operating_system
+  has_one :hardware
```
Needs `dependent => destroy`?

Fryguy commented on vmdb/app/models/configuration_manager_foreman.rb:
```diff
@@ -0,0 +1,5 @@
+class ConfigurationManagerForeman < ConfigurationManager
+  delegate :connection_attrs, :to => :provider
+  has_many :configured_systems, :class_name => :ConfiguredSystemForeman
```
Will these systems be archived when the config manager is removed?  or should the system be dependent destroyed?

Fryguy commented on vmdb/app/models/configuration_manager_foreman.rb:
```diff
@@ -0,0 +1,5 @@
+class ConfigurationManagerForeman < ConfigurationManager
+  delegate :connection_attrs, :to => :provider
+  has_many :configured_systems, :class_name => :ConfiguredSystemForeman
+  has_many :configuration_profiles
```
Same question (I'm thinking destroyed for this one).

brandondunne commented:
This is looking great! :+1:

brandondunne commented:
@Fryguy Please review

Fryguy commented on vmdb/app/models/provider.rb:
```diff
@@ -0,0 +1,18 @@
+          when false then OpenSSL::SSL::VERIFY_NONE
+          else val
+          end
+    write_attribute(:verify_ssl, val)
```
If this is a column, prefer calling `super` to `write_attribute`

Fryguy replied:
What does this actually write to the database, then?  Is VERIFY_PEER a number?

kbrock replied:
verify_peer is a number. think it is 0

Fryguy replied:
Should we make the column boolean, or is the goal to store the specific value to allow for other choices?

Fryguy commented on vmdb/app/models/provider_foreman.rb:
```diff
@@ -0,0 +1,18 @@
+  def before_save
+    build_provisioning_manager unless provisioning_manager
+    build_configuration_manager unless configuration_manager
+  end
```
Prefer using [inheritable callbacks vs override](http://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html#module-ActiveRecord::Callbacks-label-Inheritable+callback+queues)

```ruby
  before_save :init_managers

  private

  def init_managers
    build_provisioning_manager unless provisioning_manager
    build_configuration_manager unless configuration_manager
  end
```

Also, should this be on before_create instead of before_save?  The latter will be called much more frequently.

Fryguy commented on vmdb/app/models/provider_foreman.rb:
```diff
@@ -0,0 +1,18 @@
+    build_configuration_manager unless configuration_manager
+  end
+
+  def connection_attrs
```
Can you take an auth_type parameter here which defaults to nil or maybe an options hash, and then pass auth_type to authentication_userid and authentication_password?  If we will reuse this method in the future, then it's required by most of the other .connect methods, for example with [Amazon](https://github.com/ManageIQ/manageiq/blob/master/vmdb/app/models/ems_amazon.rb#L37-L44).  In addition, I wonder if credential validity should be checked here as well as in that link.

Fryguy commented on vmdb/db/migrate/20141029005602_create_configuration_managers.rb:
```diff
@@ -0,0 +1,75 @@
+      t.string     :description
+      t.belongs_to :operating_system_flavor, :type => :bigint
+      t.belongs_to :configuration_manager,   :type => :bigint
+      t.string     :manager_ref
```
nice...we've had ems_ref for so long I'm just drawn to it.  This is much better as we move forward.

Fryguy commented on vmdb/db/migrate/20141029005602_create_configuration_managers.rb:
```diff
@@ -0,0 +1,75 @@
+    add_index :configuration_profiles_customization_scripts, [:configuration_profile_id, :customization_script_id],
+              :name => :index_on_configuration_profiles_customization_scripts_i1
+    add_index :configuration_profiles_customization_scripts, [:customization_script_id, :configuration_profile_id],
+              :name => :index_on_configuration_profiles_customization_scripts_i2
```
Do you need both indexes?  Which direction is more likely to be used?

Fryguy replied:
Also, you don't need to specify both values in both directions.

Example:
If configuration_profile is more likely to be queried from, then you would need indexes on
- [:configuration_profile_id, :customization_script_id]
- :customization_script_id
Queries on both columns will always use the first index.
Queries on just :configuration_profile_id will use the first index
Queries on just :customization_script_id will just use the second index.

OR

If Postgres can handle using multiple singular indexes (need to check), then you could have indexes on
- :configuration_profile_id
- :customization_script_id
Queries on both columns will always use both indexes if necessary
Queries on just :configuration_profile_id will use the first index
Queries on just :customization_script_id will just use the second index.

Fryguy replied:
More info on Postgres' handling multiple indexes: http://www.postgresql.org/docs/9.3/static/indexes-bitmap-scans.html

kbrock replied:
yea. know bitmap scans. not a fan. (it was trashing us in the queue)
I'll get rid of one of these - thanks

Fryguy commented on vmdb/db/migrate/20141029005602_create_configuration_managers.rb:
```diff
@@ -0,0 +1,75 @@
+      t.belongs_to :configuration_profile,   :type => :bigint
+      t.belongs_to :configuration_manager,   :type => :bigint
+      t.string     :manager_ref
+      t.string     :type
```
Minor, but can we move type to the top?  It's nicer to see the STI in DB visualization closer to the front of the columns.

Fryguy commented on vmdb/db/migrate/20141029005602_create_configuration_managers.rb:
```diff
@@ -0,0 +1,75 @@
+    add_index :configured_systems_customization_scripts, [:configured_system_id, :customization_script_id],
+              :name => :index_on_configured_systems_customization_scripts_i1
+    add_index :configured_systems_customization_scripts, [:customization_script_id, :configured_system_id],
+              :name => :index_on_configured_systems_customization_scripts_i2
```
Same on if the indexes are both needed.

Fryguy commented on vmdb/app/models/computer_system.rb:
```diff
@@ -0,0 +1,6 @@
+class ComputerSystem < ActiveRecord::Base
+  belongs_to :configured_system
```
The migration has `belongs_to :managed_entity, :polymorphic => true`, but this is different.

Fryguy commented on vmdb/db/migrate/20141029005620_create_computer_systems.rb:
```diff
@@ -0,0 +1,22 @@
+      t.timestamps
+    end
+    add_index :computer_systems,  [:managed_entity_id, :managed_entity_type],
+              :name => :computer_systems_managed_entity_i1
```
Don't give a name, let Rails just choose it (unless it's too long, but is it?)

kbrock replied:
56 letters is too long. yes, it blew up. I'll try again

Fryguy replied:
Can you make this `:index_computer_systems_managed_entity`

Fryguy commented on vmdb/db/migrate/20141029005602_create_configuration_managers.rb:
```diff
@@ -0,0 +1,75 @@
+      t.string  :guid, :limit => 36
+      t.belongs_to :zone, :type => :bigint
+      t.timestamps
+    end
```
Should creating providers be separated into their own migration before this one?  Also I noticed that `create_table :configuration_managers` references providers, but technically they are being created after.  Although it works either way, it reads nicer if the providers are created first.

Fryguy commented on vmdb/db/migrate/20141029005602_create_configuration_managers.rb:
```diff
@@ -0,0 +1,75 @@
+  end
+
+  def down
+    drop_table :providers
```
and vice-versa, they should be dropped in reverse.

Fryguy commented on vmdb/db/migrate/20141029165319_create_provisioning_managers.rb:
```diff
@@ -0,0 +1,47 @@
+    add_index :customization_scripts_operating_system_flavors, [:operating_system_flavor_id, :customization_script_id],
+              :name => :index_on_customization_scripts_operating_system_flavors_i1
+    add_index :customization_scripts_operating_system_flavors, [:customization_script_id, :operating_system_flavor_id],
+              :name => :index_on_customization_scripts_operating_system_flavors_i2
```
Same on the indexes.

Fryguy commented on vmdb/spec/factories/provider.rb:
```diff
@@ -0,0 +1,13 @@
+FactoryGirl.define do
+  factory :foreman_provider, :class => 'ProviderForeman' do
```
Should there be a base :provider factory that this factory derives from?  Things like the authentications stuff can just live there instead.

kbrock replied:
I was thinking different providers did different authentication (maybe multi endpoints)

kbrock replied:
I think brandon changed this

brandondunne replied:
I just changed it to cleanup the rubocop comment.  I feel like all providers will implement authentications in their own way.  Once we have a couple more then we can refactor?

Fryguy commented on vmdb/spec/factories/provider.rb:
```diff
@@ -0,0 +1,13 @@
+    ignore do
+      userid   "admin"
+      password "smartvm"
+    end
```
I don't understand the need for this.  Can you just hardcode it below in the after(:build) and then you don't need the evaluator?

Fryguy commented on vmdb/spec/models/provider_foreman_spec.rb:
```diff
@@ -0,0 +1,36 @@
+describe ProviderForeman do
+  let(:provider) do
+    FactoryGirl.build(:foreman_provider,
+                      :url      => "example.com",
```
This can be defaulted in the factory as well.  In addition, if you are defaulting userid and password, can the test just use those default values instead of overriding?

kbrock replied:
since we are testing these values, I like to pass it in instead of relying upon a value in the factory.

I have set a default in that provider

Fryguy commented:
So, the habtms are interesting.  When I did a drive-by design with brandon some time back, we actually broke those up into separate associations...here's what we had come up with.  I don't see the granularity here, hence why I'm mentioning.

![screen shot 2015-01-12 at 11 51 01 am](https://cloud.githubusercontent.com/assets/52120/5706758/5715e3ba-9a51-11e4-8d4c-e2f291fdadf9.png)

- For the os flavors,

   ```ruby
   OsFlavor has_one customization_script_os
   OsFlavor has_one customization_script_os
   OsFlavor has_one customization_script_os
   OsFlavor has_one custtimzation_script_os
   OsFlavor has_one custtimzation_script_os
   OsFlavor has_one custtimzation_script_os
   ```

-

Fryguy commented:
Also, you're missing CustomizationScriptBoot and CustomizationScriptOs.  Are those not needed?

Fryguy commented on vmdb/app/models/configured_system_foreman.rb:
```diff
@@ -0,0 +1,4 @@
+class ConfiguredSystemForeman < ConfiguredSystem
+  belongs_to :operating_system_flavor
+  has_and_belongs_to_many :customization_scripts
```
This is interesting...It wasn't in the drive-by design.  Is it needed?  Where would this information be used.  My only concern is that if it's for genealogy purposes it would get stale.  That is, if a script is changed, then this would point to the "new" script, which would make the relationship invalid or inaccurate.  You would have to keep a whole history of scripts, which can become a huge maintenance hassle.

brandondunne replied:
I don't think we need this relationship

Fryguy commented on vmdb/app/models/configured_system.rb:
```diff
@@ -0,0 +1,7 @@
+class ConfiguredSystem < ActiveRecord::Base
+  include NewWithTypeStiMixin
+  belongs_to :configuration_manager
+  belongs_to :configuration_profile
```
Same [comment](https://github.com/ManageIQ/manageiq/pull/947/files#r22804522) goes for this relationship.

brandondunne replied:
The refresh should update this relationship.

Fryguy commented on vmdb/app/models/provider.rb:
```diff
@@ -0,0 +1,18 @@
+  def verify_ssl=(val)
+    val = case val
+          when true then OpenSSL::SSL::VERIFY_PEER
+          when false then OpenSSL::SSL::VERIFY_NONE
```
Would it make sense to store the a string or the string constant name in the database instead of the integer?  At least when you look at a database you would see the string and know the value instead of just seeing "0" and have to go figure out what that means.

brandondunne replied:
I think it's clear enough to store the integer.

Fryguy commented on vmdb/app/models/provider.rb:
```diff
@@ -0,0 +1,18 @@
+          when false then OpenSSL::SSL::VERIFY_NONE
+          else val
+          end
+    super(val)
```
I think you can just say `super` without the val and it will work just fine.

brandondunne replied:
This ```val``` is modified from the original one passed in.

kbrock replied:
the value of `val` changed, so I think we need to pass it

kbrock commented:
@brandondunne @Fryguy 

- renamed factory to provider_foreman
- removed habtm from computer system
- converted config profile to have 2 customization scripts rather than habtm
- removed `val` from super call in `Provider#verify_ssl=`
